### PR TITLE
docs: add Caddy+LetsEncrypt TLS example

### PR DESCRIPTION
**Pull request by bpmct**
This is inspired by [GitLab's Caddy Recipe](https://gitlab.com/gitlab-org/gitlab-recipes/blob/master/web-server/caddy/README.md) and I plan to add additional ones for NGINX and Kubernetes ingress+cert-manager. 

As much as possible, I tried to reference the Caddy docs for custom builds/providers, while showing one concrete example for wildcards: AWS Route53. If we want to make it slightly more concrete, I could actually modify the docker-compose to support building a custom Dockerfile there that the user could just uncomment.

I understand this adds some debt to the codebase, so I'm open to other ideas. I could just include code snippets in the docs but I found that structure a bit confusing. One advantage of having this in the codebase is I could create an automated (weekly) test to spin up a Coder server with Caddy and ensure everything works as planned. If we wanted to do that as a prerequisite to merging, I'd totally understand.

Contributes to #3518

### CODE REVIEW

File: examples/web-server/caddy/README.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+   # Enable UncomplicatedFirewall
+   sudo ufw enable
+   ```
+
+1. Navigate to your Coder URL! A TLS certificate should be auto-generated on your first visit.
+
+## Generating wildcard certificates
+
+By default, this configuration uses Caddy's [on-demand TLS](https://caddyserver.com/docs/caddyfile/options#on-demand-tls) to generate a certificate for each subdomain (e.g. `app1.coder.example.com`, `app2.coder.example.com`). When users visit new subdomains, such as accessing [ports on a workspace](../../networking/port-forwarding.md), the request will take an additional 5-30 seconds since a new certificate is being generated.
```
Comment (bpmct) [inline examples/web-server/caddy/README.md:97]: > the request will take an additional 5-30 seconds

I'm fairly certain this is the only downside of using Caddy's on-demand TLS. While LetsEncrypt's rate limit (50 certificates/domain/week) can quickly be exhausted, Caddy will silently fall back to ZeroSSL which has no rate limit. ZeroSSL is significantly slower though. If you think this needs further explanation in the docs, I can expand.

With that being said, the extra effort for a wildcard is worth it if a Coder deployment is being actively being used in production.

File: examples/web-server/caddy/docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,57 @@
+version: "3.9"
```
Comment (ghuntley) [inline examples/web-server/caddy/docker-compose.yaml:1]: What is the purpose of `~/docker-compose.yaml` vs `~/examples/web-server/caddy/docker-compose.yaml`? 

ie. does it make sense to delete one?

Comment (bpmct) [reply examples/web-server/caddy/docker-compose.yaml:1]: `~/docker-compose.yaml` uses the tunnel by default which means that Caddy + a domain name is not required. However, I think we could move the compose to `~/examples/install/docker` or something.

Comment (ghuntley) [reply examples/web-server/caddy/docker-compose.yaml:1]: 👍 re:  I think we could move the compose to ~/examples/install/docker or something in an additional PR.

**Review by ammario — COMMENTED** [review]
I have no experience with Caddy so can't give an approval here. Looks cool though.

**Review by sharkymark — APPROVED** [review]

**Pull request merged**